aten/src/ATen/native/mkldnn/xpu/detail/WoQMatmul.cpp

```diff
@@ -0,0 +1,183 @@
+#include <c10/xpu/XPUFunctions.h>
+
+#include <ATen/native/mkldnn/xpu/detail/Attr.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+
+#include <oneapi/dnnl/dnnl.hpp>
+#include <cstdint>
+
+
+namespace at::native::onednn {
+
+sycl::event woq_matmul_int4(
+    Tensor& result, //torchao: [M, K]
+    const Tensor& mat1_, //torchao: [M, K]
+    const Tensor& mat2_, //torchao quantized weight, [K/8, N]
+    const Tensor& scale, //torchao: [K/group_size, N]
+    const Tensor& zp, //torchao: [K/group_size, N]
+    int64_t group_size,
+    Attr attr,
+    const std::vector<sycl::event>& deps) {
+  size_t dims = result.dim();
+  TORCH_CHECK(
+      dims == 2 || dims == 3,
+      "oneDNN matmul only works with 2D or 3D, got ",
+      dims);
+  TORCH_CHECK(result.defined(), "oneDNN matmul result should be defined");
+
+  at::Device cur_device = at::Device(at::kXPU, at::xpu::current_device());
+  auto engine = GpuEngineManager::Instance().get_engine(cur_device);
+  auto engine_index = cur_device.index();
+  auto stream = GpuStreamManager::Instance().get_stream();
+
+  // make them all contiguous
+  Tensor m1 = is_onednn_matmul_strides(mat1_) ? mat1_ : mat1_.contiguous();
+  Tensor m2 = is_onednn_matmul_strides(mat2_) ? mat2_ : mat2_.contiguous();
+  Tensor scale_ = is_onednn_matmul_strides(scale) ? scale : scale.contiguous();
+  Tensor zp_ = is_onednn_matmul_strides(zp) ? zp : zp.contiguous();
+  Tensor dst = is_onednn_matmul_strides(result, true) ? result : result.contiguous();
+  int m = m1.size(-2); // M
+  int n = dst.size(-1);  //m2.size(0) * kNTileSize; 
+  int k = m1.size(-1); // K1
+  int64_t mb = 1;
+
+  // Construct usr md from input
+  // xxx_usr_md would describe the real layout of inputs
+  auto m1_usr_dt = get_onednn_dtype(m1); // e.g., half <==> f16
+  auto m2_usr_dt = get_onednn_dtype(m2); //int32 tensor, pack 8 int4
+  auto scale_usr_dt = get_onednn_dtype(scale_); // bf16
+  //   auto zp_usr_dt = dnnl::memory::data_type::s4; // int32, representing 8xint4
+  auto zp_usr_dt = get_onednn_dtype(zp_); // bf16
+  auto dst_usr_dt = get_onednn_dtype(dst); // bf16
+
+  dnnl::memory::dims m1_usr_dims, m2_usr_dims, scale_usr_dims, zp_usr_dims, dst_usr_dims;
+  dnnl::memory::dims m1_usr_strides, m2_usr_strides, scale_usr_strides, zp_usr_strides, dst_usr_strides;
+  const uint64_t compressed_k = (uint64_t)(k / 8);
+  const uint64_t compressed_n = (uint64_t)(n / 8);
+  const uint64_t num_groups = (uint64_t)(k / group_size); 
+    //wei: {compressed_k, n}:   shape:{n/8, 32, xxx , innerKTiles/2}
```

Comment:
weight is not fake tensor for xpu and the the shape is {compressed_k, n}.

Reply:
thanks for pointing these out, I have modified the code here, thanks.